datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restoreindices/SendMAEStep.java

```diff
@@ -111,42 +122,185 @@ public Function<UpgradeContext, UpgradeStepResult> executable() {
+            ignored = ignored + 1;
             continue;
           }
 
           SystemMetadata latestSystemMetadata = EntityUtils.parseSystemMetadata(aspect.getSystemMetadata());
 
           // 5. Produce MAE events for the aspect record
           _entityService.produceMetadataChangeLog(urn, entityName, aspectName, aspectSpec, null, aspectRecord, null,
-              latestSystemMetadata,
-              new AuditStamp().setActor(UrnUtils.getUrn(SYSTEM_ACTOR)).setTime(System.currentTimeMillis()),
-              ChangeType.RESTATE);
+                  latestSystemMetadata,
+                  new AuditStamp().setActor(UrnUtils.getUrn(SYSTEM_ACTOR)).setTime(System.currentTimeMillis()),
+                  ChangeType.RESTATE);
 
-          totalRowsMigrated++;
+          rowsMigrated++;
         }
-        context.report().addLine(String.format("Successfully sent MAEs for %s rows", totalRowsMigrated));
-        start = start + count;
+
         try {
-          TimeUnit.MILLISECONDS.sleep(getBatchDelayMs(context.parsedArgs()));
+          TimeUnit.MILLISECONDS.sleep(batchDelayMs);
         } catch (InterruptedException e) {
           throw new RuntimeException("Thread interrupted while sleeping after successful batch migration.");
         }
+        return new KafkaJobResult(ignored, rowsMigrated);
       }
+  }
+
+  public SendMAEStep(final EbeanServer server, final EntityService entityService, final EntityRegistry entityRegistry) {
+    _server = server;
+    _entityService = entityService;
+    _entityRegistry = entityRegistry;
+  }
+
+  @Override
+  public String id() {
+    return "SendMAEStep";
+  }
+
+  @Override
+  public int retryCount() {
+    return 0;
+  }
+
+  private KafkaJobResult iterateFutures(List<Future<KafkaJobResult>> futures) {
+    int beforeSize = futures.size();
+    int afterSize = futures.size();
+    while (beforeSize == afterSize) {
+      try {
+        TimeUnit.SECONDS.sleep(1);
+      } catch (InterruptedException e) {
+        // suppress
+      }
+      for (Future<KafkaJobResult> future: new ArrayList<>(futures)) {
+        if (future.isDone()) {
+          try {
+            KafkaJobResult result = future.get();
+            futures.remove(future);
+            return result;
+          } catch (InterruptedException | ExecutionException e) {
+            e.printStackTrace();
+          }
+        }
+      }
+      afterSize = futures.size();
+    }
+    return null;
+  }
+
+  @Override
+  public Function<UpgradeContext, UpgradeStepResult> executable() {
+    return (context) -> {
+
+      long startTime = System.currentTimeMillis();
+
+      int batchSize = getBatchSize(context.parsedArgs());
+      int numThreads = getThreadCount(context.parsedArgs());
+      long batchDelayMs = getBatchDelayMs(context.parsedArgs());
+      Optional<String> aspectName;
+      if (containsKey(context.parsedArgs(), RestoreIndices.ASPECT_NAME_ARG_NAME)) {
```

Comment:
I think we should move away from the command line arguments approach, it's less flexible than just pure ENV and requires us to modify the helm chart templates every time we add new configuration.